libdnf::Transaction::run() method and rpm transaction lock

m-blaha opened the pull request.

m-blaha committed:
Lock files class

m-blaha committed:
dnfdaemon: Use libdnf method to run rpm transaction

jrohel commented on libdnf/utils/locker.cpp:
```diff
@@ -0,0 +1,63 @@
+
+    // create a local unix socket
+    socket_fd = socket(AF_UNIX, SOCK_DGRAM, 0);
+    libdnf_assert(socket_fd > 0, "Cannot create a lock - socket creation failed.");
```
Replace `socket_fd > 0` with `socket_fd != INVALID_FD`.

lukash replied:
Well more importantly, this should not be an assert at all, the result of the `socket()` call is a run-time error. That error has an `errno` which should be part of the message. Please pay attention to error handling...

jrohel commented on libdnf/utils/locker.hpp:
```diff
@@ -0,0 +1,41 @@
+
+namespace libdnf::utils {
+
+constexpr int INVALID_FD = -1;
```
Do we want `INVALID_FD` to be public? What to define it as a private member of the Locker class?

lukash replied:
Tbh. I don't see a reason to have this in a variable, just use a -1 directly?

jrohel replied:
This is a special descriptor value expressing an invalid descriptor. Naming doesn't hurt anything. Of course definitely not a variable. But there is `constexpr` in the code. I have no problem with that.

lukash replied:
Sure... There's also an invalid descriptor expressed by a `-1` here: https://github.com/rpm-software-management/libdnf/blob/2f528454f867f409e28300f0c4a3d93f2fabac18/libdnf/utils/temp.hpp#L88

Shall we share the `INVALID_FD` definition to deduplicate? If not, I want to approach this consistently, so there should be an `INVALID_FD` constant in that file as well. Ultimately, `-1` is just simpler, it saves on LoCs, and there's no other difference.

jrohel commented on libdnf/utils/locker.cpp:
```diff
@@ -0,0 +1,63 @@
+    struct sockaddr_un lock_address;
+    memset(&lock_address, 0, sizeof(lock_address));
+    lock_address.sun_family = AF_UNIX;
+    // first byte of abstract socket address is required to be \0
```
The abstract socket namespace is a non portable Linux extension.
What to use something else? Pathname socket? Lock file?

m-blaha replied:
Right, it will work only on linux. What I liked on this solution is independence on any filesystem path. You just use arbitrary name for the lock. But in case we need compatibility across different operating systems, I would go with a lock file.

jrohel commented on libdnf/utils/locker.cpp:
```diff
@@ -0,0 +1,63 @@
+namespace libdnf::utils {
+
+Locker::Locker(const std::string & name) {
+    // unix domain socket path is declared as char[108]
```
Using this magic constant seems dangerous.
Yes 108 on Linux. But only 104 on FreeBSD.
How about using real value instead of a magic number?
```
constexpr int max_path_len = (sizeof sockaddr_un::sun_path) - 1;
libdnf_assert(name.size() <= max_path_len, "Lock name is too large (maximum length {} expected)", max_path_len);
```

jrohel commented on libdnf/utils/locker.cpp:
```diff
@@ -0,0 +1,63 @@
+
+Locker::Locker(const std::string & name) {
+    // unix domain socket path is declared as char[108]
+    libdnf_assert(name.size() <= 107, "Lock name is too large (maximum length 107 expected)");
```
Using this magic number seems dangerous.
Yes `108` on Linux. But only `104` on FreeBSD.
How about using real value instead of a magic number?
```
constexpr int max_path_len = (sizeof sockaddr_un::sun_path) - 1;
libdnf_assert(name.size() <= max_path_len, "Lock name is too large (maximum length {} expected)", max_path_len);
```

lukash replied:
I don't think this is a case for an assert either, is it? The path can ultimately be influenced by user input, like in a configuration, maybe it's inside an installroot, right? In that case we don't want a crash, this situation can occur normally.

m-blaha replied:
I'm not sure here about the assert. It's not supposed to be a public API so users should not call this directly. The `path` in this case is just arbitrary identifier of the lock, it's not a file path directly influenced by installroot. Before calling a lock we need to construct this identifier and make sure it's length is shorter than the maximum.

j-mracek commented on include/libdnf/base/transaction.hpp:
```diff
@@ -64,6 +74,33 @@ class Transaction {
+    std::optional<std::string> get_cmdline() const noexcept;
+
+    /// Set command line of a program that started the transaction
+    void set_cmdline(const std::string & value);
```
In case I am using GUI (Gnome Software), what will be expected to be stored here?

j-mracek replied:
I guess that we use `cmdline` as a label of particular transaction. Mostly it contains cmdline command.

j-mracek commented on include/libdnf/base/transaction.hpp:
```diff
@@ -64,6 +74,33 @@ class Transaction {
+    void set_user_id(uint32_t value);
+
+    /// Prepare, check and run rpm transaction
+    TransactionRunResult run(libdnf::rpm::TransactionCallbacks & callbacks);
```
Please correct me if I am wrong, `comment` is optional argument, `user_id` overrides default detection (Used by daemon), and `cmdline` is more less like required argument. We want to have all transaction with some label provided by API user, microdnf, or Daemon.
What about to use run(libdnf::rpm::TransactionCallbacks & callbacks, const std::string & cmdline) + two optional arguments?

m-blaha replied:
Well, it's not entirely clear what the `cmdline` should be for dbus API (or basically any API) use-case. All you can put there is just generic `<path to dnfdaemon-server>`.

j-mracek replied:
I think it will be nice if we will have no transaction without a description who set it up. For gui/daemon tools it will be enough to write down the name of the client, if it will be possible (Gnome Software, ...). What do you think?

m-blaha replied:
OK, makes sense. I've removed the attributes from `Transaction` class and changed them to arguments of `run()` method.

j-mracek commented on include/libdnf/common/proc.hpp:
```diff
@@ -33,6 +33,12 @@ constexpr uid_t INVALID_UID = static_cast<uid_t>(-1);
+/// The getuid() is returned instead if there was a problem.
+/// The value is cached.
+/// @since 5.0
+uid_t get_login_uid() noexcept;
```
Do we need it in public API?

j-mracek replied:
I guess Daemon client requires it.

m-blaha replied:
Not the client, but the server will need probably both functions declared here - read_login_uid_from_proc() as primary source of uid and get_login_uid() as a fall back.

jrohel commented on include/libdnf/base/transaction.hpp:
```diff
@@ -64,6 +74,21 @@ class Transaction {
+    /// @param user_id      UID of the user that started the transaction
+    /// @param comment      any comment describing the transaction
+    /// @return libdnf::base::Transaction::TransactionRunResult
+    TransactionRunResult run(libdnf::rpm::TransactionCallbacks & callbacks, const std::string & cmdline, const std::optional<uint32_t> user_id, const std::optional<std::string> comment);
```
Too long line. Try the clang-format.

jrohel replied:
Here we have a problem with API abstractions.
The `run` method requires "rpm" specific callbacks. Even according to the description, is "rpm" specific: "Prepare, check and run **rpm transaction**".
What in the future? Will we have eg `run_deb`? Or another `run` with an overloaded callbacks argument.
Maybe add TODO. But I'm thinking a lot ahead.

jrohel commented on libdnf/base/transaction_impl.hpp:
```diff
@@ -62,10 +63,13 @@ class Transaction::Impl {
         const std::set<std::string> & additional_data,
         bool strict);
 
+    TransactionRunResult run(libdnf::rpm::TransactionCallbacks & callbacks, const std::string & cmdline, const std::optional<uint32_t> user_id, const std::optional<std::string> comment);
```
Too long line. Try the clang-format.

jrohel commented on microdnf/utils.hpp:
```diff
@@ -30,11 +30,6 @@ namespace microdnf {
-/// Gets the login uid, if available.
-/// The getuid() is returned instead if there was a problem.
-/// The value is cached.
-uid_t get_login_uid() noexcept;
```
You copied the `get_login_uid` function to libdnf in the previous commit. And now you've just deleted the header. Please also remove the implementation from "utils.cpp".

m-blaha committed:
Transaction::run() method

This method creates and executes an rpm transaction, updates the system
state and writes record to the history database.

m-blaha committed:
microdnf: Use libdnf method to run rpm transaction

jrohel commented:
Thanks.

lukash commented on libdnf/utils/locker.cpp:
```diff
@@ -0,0 +1,61 @@
+Locker::LockResult Locker::lock() {
+    lock_fd = open(path.c_str(), O_CREAT | O_RDWR, 0660);
+    if (lock_fd == -1) {
+        return LockResult::ERROR_FD;
```
This just hides any error that occurred during the `open()` call behind an opaque `ERROR_FD` value.

lukash commented on libdnf/utils/locker.cpp:
```diff
@@ -0,0 +1,61 @@
+    fl.l_pid = 0;
+    auto rc = fcntl(lock_fd, F_SETLK, &fl);
+    if (rc == -1) {
+        return LockResult::ERROR_LOCK;
```
While, looking at the `fcntl()` man page, any other error besides "already locked" seems unlikely (assuming root privileges), we should only return "locked" state (and note it's not an actual error as it's labeled here) for that particular `errno` by which it's represented and throw an error otherwise.

lukash commented on libdnf/utils/locker.cpp:
```diff
@@ -0,0 +1,61 @@
+void Locker::unlock() {
+    if (lock_fd != -1) {
+        unlink(path.c_str());
+        close(lock_fd);
```
It's also possible for these to fail and therefore for `unlock()` as a whole, the errors should also be handled here.

lukash commented on libdnf/base/transaction.cpp:
```diff
@@ -281,5 +319,108 @@ TransactionPackage Transaction::Impl::make_transaction_package(
+    std::filesystem::path lock_file_path = system_cache_dir;
+    lock_file_path /= "rpmtransaction.lock";
+    libdnf::utils::Locker locker(lock_file_path);
+    if (locker.lock() != libdnf::utils::Locker::LockResult::SUCCESS) {
```
... and this just erases any distinction between various errors that can occur and simply fails with an "Already locked" message. Any error working with the lock will be hidden from the user and they'll be left clueless. Note all kinds of weird stuff may happen in e.g. containers and non-standard environments. This has to be fixed, also for our own good, because we'll be getting the bug reports that won't contain any useful information.

lukash reviewed (commented):
Sorry about being late here, I meant to check this sooner :slightly_frowning_face:
